### ISSUE TITLE
Why is plugin-import is peerDep of node-resolver?

### ISSUE THREAD

**Issue by jasonkarns** [benmosher/eslint-plugin-import#437]
I am tracking down some very strange shrinkwrap issues caused by eslint/eslint-plugin-import and I came across this:

https://github.com/benmosher/eslint-plugin-import/blob/71afde73b8d18a487671d1d4ae0534c0ca52e938/resolvers/node/package.json#L31

Why is eslint-plugin-import (hereafter "import") listed as a peerDep of eslint-import-resolver-node (hereafter "resolver")? Is my understanding correct that in no scenario does resolver expect to be a sibling of import?

Resolver is already a direct dependency of import. And resolver does not `require` or rely on any of the functionality of import to function.

**Comment by benmosher** [comment]
https://nodejs.org/en/blog/npm/peer-dependencies/#the-problem-plugins

**TL;DR:** enforcing that the `resolver` is used only with versions of `import` that should work well together.

That post is fairly dated, though. Is there a better (or maybe just _newer_?) way to ensure compatibility?

**Comment by jasonkarns** [comment]
`resolver` isn't a plugin for (nor a peer of) `import`, it's a direct dep. As such, the way to ensure compatibility is already being done: `import` specifies precisely which version(s) of `resolver` it requires:  https://github.com/benmosher/eslint-plugin-import/blob/71afde73b8d18a487671d1d4ae0534c0ca52e938/package.json#L75

`peerDep` is only to help ensure compatibility of modules that will be _peers_ (hence the name) which is only useful for modules that are plugins for (installed as siblings of) the respective peerDep. (like eslint-plugin-import is installed as a peer of eslint)

**Comment by jasonkarns** [comment]
I should also note that, because npm2 actually _installs_ peerDependencies as siblings (rather than simply emitting warnings as npm3 does), it is common with `resolve` having `import` as a peerDep, npm will actually install `import` as a sibling to `resolve`. And since `resolve` is a direct dep of `import`, that means the final node_modules on disk have `import` as a dependency *of itself* (within its own node_modules).

**Comment by jasonkarns** [comment]
I've been able to boil this down to a minimal reproducible issue, that is actually causing production problems for us: (only manifests on npm2)

`npm init -f; npm install --save eslint-plugin-import@1.8.0 eslint@^2.9.0; npm shrinkwrap; rm -r node_modules; npm install; npm shrinkwrap; rm -r node_modules; npm install; npm shrinkwrap; rm -r node_modules; npm install;`

The above demonstrates a scratch package with direct deps installed: `eslint ^2.9.0` and `eslint-plugin-import@1.8.0`

Given such a package, running the following commands: `npm shrinkwrap; rm -r node_modules; npm install` will fail on the third install due to versioning conflicts of eslint:

```
npm ERR! install trying to install 3.0.1 to /private/tmp/take2/node_modules/eslint
npm ERR! install but already installed versions [ '2.13.1' ]
... [100s of lines snipped]
```

**Comment by benmosher** [comment]
Ah, wild. There were a handful of issues that I never understood around this. 

The idea originally was that all the resolvers would be plugins, but then I bundled the Node one because I figured that it covered most use cases. My hope was that it could still be installed as a peer to choose a different version than the bundled one. 

I don't think that actually makes sense, as you've clearly stated. Thanks!

@jfmengels: feels like a crossroads between making the node resolver fully optional vs removing the peer dependency. What do you think?

I'm tempted to stop bundling it and instead to have the recommended config reference it, but then it's one more step for the average user. So I'm not sure which direction is more practical.

**Comment by jasonkarns** [comment]
Personally I think the optimal ending scenario would be having the node-resolver still included as a direct dependency as it is now, while allowing end users to provide another resolver of their choosing via configuration. (I would model this how most test frameworks/linters etc have bundled or direct-dep reporters, but still allow custom reporters to be referenced via config.)

However, and I think this is the critical part, regardless of whether the node-resolver is left as a direct-dep or fully extracted/made optional; the peerDep listing is having no effect other than to cause plugin-import to be installed into its own node_modules. The removal of the peerDep line from the node resolver can be done without any other changes and the final decision on bundling/configuring can be made separately.

If you decide to go with having end-users choose the resolvers, then it will need extracted and at *that* point the peerDep line would be appropriate. But until then, it is adding no value while blocking shrinkwrap installs.

**Comment by benmosher** [comment]
Fair enough. I will get a patch out this morning.

**Comment by jfmengels** [comment]
> Personally I think the optimal ending scenario would be having the node-resolver still included as a direct dependency as it is now, while allowing end users to provide another resolver of their choosing via configuration

This sounds good to me

**Comment by benmosher** [comment]
Yeah, I think that's right.

**Comment by benmosher** [comment]
Just published v0.2.2 of Node resolver to npm, with #438 merged. Should be compatible back to roughly v1.4 of the import plugin.

Thanks for this, @jasonkarns! 😁

**Comment by jasonkarns** [comment]
Thank you @benmosher!
